### PR TITLE
Use actual etcd client for /healthz/etcd checks

### LINKED ISSUE #64909: etcd healthcheck is too simple
currently, the etcd healthcheck is too simple. It only checks tcp connection https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/storage/etcd3/preflight/checks.go#L65. 

I think we can enhance it the way etcdctl does. 
etcdV2:  Get(url + "/health")
https://github.com/coreos/etcd/blob/6e39a39e3a0fb4258ed65e781e5d65849ee25d48/etcdctl/ctlv2/command/cluster_health.go#L83

etcdv3: cli.Get(ctx, "health")
https://github.com/coreos/etcd/blob/ff455e3567d6465717d3ced6dc889d4eec9ed890/etcdctl/ctlv3/command/ep_command.go#L121

/sig api-machinery

### PR DESCRIPTION
**Pull request by liggitt**
* avoids redialing etcd on every health check (which makes slow DNS a false-positive healthz failure)
* ensures etcd TLS setup is correct (errors verifying the etcd API or sending client credentials manifest as healthz failures)
* ensures the etcd cluster is actually responsive

fixes #64909

```release-note
Etcd health checks by the apiserver now ensure the apiserver can connect to and exercise the etcd API
```

### CODE REVIEW

**Comment by liggitt** [comment]
@kubernetes/sig-api-machinery-pr-reviews 

/cc @xiang90 @gyuho

File: staging/src/k8s.io/apiserver/pkg/server/options/etcd.go [hunk 1]
```diff
@@ -181,29 +181,30 @@ func (s *EtcdOptions) ApplyTo(c *server.Config) error {
...
 }
 
-func (s *EtcdOptions) addEtcdHealthEndpoint(c *server.Config) {
+func (s *EtcdOptions) addEtcdHealthEndpoint(c *server.Config) error {
+	healthCheck, err := storagefactory.CreateHealthCheck(s.StorageConfig)
+	if err != nil {
+		return err
+	}
 	c.HealthzChecks = append(c.HealthzChecks, healthz.NamedCheck("etcd", func(r *http.Request) error {
-		done, err := preflight.EtcdConnection{ServerList: s.StorageConfig.ServerList}.CheckEtcdServers()
```
Comment (xiang90) [inline staging/src/k8s.io/apiserver/pkg/server/options/etcd.go]: Is the checkEtcdServers func used by anything else after this PR?

File: staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go [hunk 2]
```diff
@@ -38,15 +41,49 @@ var (
...
+		clientErrMsg.Store("")
+		return true, nil
+	}, wait.NeverStop)
+
+	return func() error {
+		if errMsg := clientErrMsg.Load().(string); len(errMsg) > 0 {
+			return fmt.Errorf(errMsg)
+		}
+		client := clientValue.Load().(*clientv3.Client)
+		ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go]: hrmm, in testing, this timeout isn't firing as expected... taking down the etcd server makes this hang indefinitely

File: staging/src/k8s.io/apiserver/pkg/server/options/etcd.go (hunk 1 above)
Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/server/options/etcd.go]: it looks like it's used as a gate when starting cmd/kube-apiserver... the etcd endpoint must be able to establish a tcp connection, but that's it. if we want to remove that pre-check, and delete the package entirely, I'd like to do it in a follow up.

Comment (xiang90) [reply staging/src/k8s.io/apiserver/pkg/server/options/etcd.go]: sure. thanks.

**Comment by liggitt** [comment]
tested this with both etcd2 and etcd3. when the etcd server was unavailable, /healthz/etcd returned failure messages without hanging. when the etcd server became available again, the /healthz/etcd endpoint recovered and started returning success again

File: staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go [hunk 3]
```diff
@@ -38,15 +41,49 @@ var (
...
 )
 
-func newETCD3Storage(c storagebackend.Config) (storage.Interface, DestroyFunc, error) {
+func newETCD3HealthCheck(c storagebackend.Config) (func() error, error) {
+	// constructing the etcd v3 client blocks and times out if etcd is not available.
+	// retry in a loop until we successfully create the client, storing the client or error encountered
+
+	clientValue := &atomic.Value{}
+
+	clientErrMsg := &atomic.Value{}
```
Comment (hzxuzhonghu) [inline staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go]: what's this for? I think If wait.PollUntil is not blocking, this is useful.

File: staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go [hunk 4]
```diff
@@ -38,15 +41,49 @@ var (
 	dialTimeout = 10 * time.Second
 )
 
-func newETCD3Storage(c storagebackend.Config) (storage.Interface, DestroyFunc, error) {
+func newETCD3HealthCheck(c storagebackend.Config) (func() error, error) {
+	// constructing the etcd v3 client blocks and times out if etcd is not available.
+	// retry in a loop until we successfully create the client, storing the client or error encountered
+
+	clientValue := &atomic.Value{}
```
Comment (hzxuzhonghu) [inline staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go]: ditto

File: staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go (hunk 3 above)
Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go]: > I think If wait.PollUntil is not blocking, this is useful.

good catch, in my tests, etcd was always available. backgrounded the client init loop as I intended to originally

**Comment by jpbetz** [comment]
cc @cheftako

**Comment by lavalamp** [comment]
@jpbetz @liggitt If healthz fails, kubelet will restart apiserver. Is restarting apiserver a good response to etcd health check failing? I suspect it will only make things worse? We need a readiness endpoint and not just liveness...

**Comment by liggitt** [comment]
> Is restarting apiserver a good response to etcd health check failing? I suspect it will only make things worse?

not necessarily... it fixes stuck tcp connections, which we have observed happening in failover cases

**Comment by lavalamp** [comment]
*time passes*

It does seem like a good idea to drop all watches if our etcd goes away, and a restart is an especially (excessively?) thorough way to accomplish that...

However! It doesn't seem great for apiserver to continuously fail liveness checks and crash-loop while etcd is down / split; it should be sufficient to just not be ready. So, maybe liveness shouldn't care about etcd until we observe a healthy etcd and become ready, at which point if we lose readiness because etcd goes away, we should also lose liveness and be restarted.

We can probably come up with something better / more principled than that: AFAIK, no one has payed any holistic attention to apiserver's liveness / readiness checks.

**Pull request merged**